### PR TITLE
Add WithVolumeMountForData

### LINKED ISSUE #837: Make it easier to use external volumes for database containers
I found my way around creating a database container with a volume mount

```csharp
var database = builder.AddPostgresContainer("database")
    .WithVolumeMount("../../data/database/data", "/var/lib/postgresql/data");
```

It would be cool to have `.WithVolumeMountForData("../../data/database/data")` offered as an extension method to `IResourceBuilder<PostgresContainerResource>`

It would be even cooler if the libraries would push us towards the pit of success by assuming that we want to use a volume mount for the database data.

### PR DESCRIPTION
**Pull request by stbau04**
Make it easier to use external volumes for database containers:
- Added WithVolumeFor.../WithBindMountFor... for data, logs and init folders
  - For the data directory on all containers
  - For the logs directory on containers with logs enabled by default (as it might confuse if the WithVolumeForLogs is available but log files are not actually enabled in the container configuration)
  - For the init folder where a /docker-entrypoint-initdb.d folder is avilable (did not find something similar on the other containers)

Fixes https://github.com/dotnet/aspire/issues/837
 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/1317)

### CODE REVIEW

File: src/Aspire.Hosting/Postgres/PostgresBuilderExtensions.cs [hunk 1]
```diff
@@ -62,6 +62,19 @@ public static IResourceBuilder<PostgresDatabaseResource> AddDatabase(this IResou
...
 
+    /// <summary>
+    /// Adds a data volume mount to a PostgreSQL container resource (mounted to /var/lib/postgresql/data in the container).
+    /// </summary>
+    /// <typeparam name="T">The resource type.</typeparam>
+    /// <param name="builder">The resource builder.</param>
+    /// <param name="source">The source path of the volume. This is the physical location on the host.</param>
+    /// <param name="type">The type of volume mount.</param>
+    /// <returns>The <see cref="IResourceBuilder{T}"/>.</returns>
+    public static IResourceBuilder<T> WithVolumeMountForData<T>(this IResourceBuilder<T> builder, string source, VolumeMountType type = default) where T : PostgresContainerResource
```
Comment (stbau04) [inline src/Aspire.Hosting/Postgres/PostgresBuilderExtensions.cs:73]: Should the isReadonly-Parameter from WithVolumeMount also be available for this method? Are there any common cases where this could be usefull?

**Comment by DamianEdwards** [comment]
Thanks for this. We should add overloads of this method that support the other database container resources too, with the appropriate path in those cases.

**Comment by stbau04** [comment]
Before adding the other overloads i would wait for the discussion on the issue (https://github.com/dotnet/aspire/issues/837#issuecomment-1850635415). If this is still the approch we stick too i'll continue adding them.

**Comment by stbau04** [comment]
@josephaw1022 created a second (duplicate) PR for this one: https://github.com/dotnet/aspire/pull/1447

**Comment by stbau04** [comment]
Probably we should wait for https://github.com/dotnet/aspire/pull/1521 and then see how it affects this pr (especially the volume/bind mount stuff)

**Comment by davidfowl** [comment]
@stbau04 Would you like to resurrect this PR?

**Comment by davidfowl** [comment]
@DamianEdwards wdyt?

**Comment by DamianEdwards** [comment]
@davidfowl yeah I like this a lot actually. @stbau04 if you're around it would be great if you could update this for changes since Dec (mainly the split out of `WithBindMount` and `WithVolume` from `WithVolumeMount`), otherwise I'm happy to take this over for preview.5

**Comment by stbau04** [comment]
@davidfowl @DamianEdwards Yes sure, i'll update it

**Comment by davidfowl** [comment]
I'm gonna assign this to you @stbau04. Thanks!

**Comment by DamianEdwards** [comment]
Seems you did a merge instead of a rebase so this PR is showing up as including **all** changes on main since you originally created it! Can you go back to before you merged and do a rebase on main instead, and then force push to your fork/branch?

**Comment by davidfowl** [comment]
@DamianEdwards  @mitchdenny Do we want to align here with the `UsePersistance` calls https://github.com/dotnet/aspire/blob/1d868e8532bdfb874e29b3ca6383b9cb9e69d0b4/src/Aspire.Hosting.Azure/Extensions/AzureStorageExtensions.cs#L123-L128?

Feels like we should have a consistent pattern.

**Comment by mitchdenny** [comment]
I'm not sure I like the interfaces in this PR. It seems to me that persistence for each container is a fairly container specific thing and we might want to have a specific extension method for each container. Perhaps in that case the overload takes a `basePath` and then if the container supports init/log/data directories then a bindmount is applied for each one relative to the base path.

**Comment by mitchdenny** [comment]
Thinking about this some more, I think what we want here is a per container type extension that takes into consideration the various quirks assocaited with each container. So instead of a pattern that would seem people doing this:

```csharp
builder.AddRedis("redis").WithBindMountForLogs(path).WithBindMountForInit(path).WithBindMountForData(path)
```

We could just do this:

```csharp
builder.AddRedis("redis").WithPersistence(path);
```

The benefit is that it is a simpler gesture for the consumer of the container and it allows us to abstract away wierdness that some containers might have. For example, the Cosmos container which we use for the emulator doesn't seem to work with bind mounts -- so this mechanism wouldn't work. Instead for that we'd need to implement a `WithPersistence(...)` method which computes a unique volume name and use that instead of a bind mount.

So in summary ... its good to make it easy for people to turn on persistence for a given container (we should do that!) but lets not make it generic when it probably isn't because we have to live with this abstraction for a while ;)

**Comment by DamianEdwards** [comment]
I think we need more than a single method if for no other reason than to support bind mounts vs. volumes. Perhaps it's just two overloads (one accepting a path and another not) but I think that's less clear than the proposed methods. Personally not a fan of the `WithPersistance` naming generally.

**Comment by DamianEdwards** [comment]
Perhaps a simple change to `WithDataVolume` and WithDataBindMount`, etc. for these. The interface approach seems fine for now I think? What kind of thing would trip that approach up?

**Comment by stbau04** [comment]
@DamianEdwards I've rebased it now
Where did you see all files included, it only showed 10 changed files to me? I've always updated my branches by merging, didn't know that this is not the correct way (or that it might show all files as changes)

**Comment by DamianEdwards** [comment]
@stbau04 looks better now thanks. It was in the GitHub PR UI. It was hanging the browser due to the number of changed files. But it's showing 10 now :smile:

**Comment by DamianEdwards** [comment]
To address @mitchdenny's concerns here I think a good compromise would be to get rid of the interfaces and just establish a pattern for these kind of container-based resources by adding methods like `WithDataVolume` and `WithDataBindMount` where they support it. If a particular container image has quirks beyond just needing to know the path to mount the volume or bind mount to, it can be handled in that method.

For init and logs, I think it only makes sense to have support for bind mounts for init, and both volume and bind mounts for logs. Reasoning being that init is all about an input to the container (run these scripts) whereas data and logs are potentially input and output.

In the case of volumes, we should support auto-generating a name based on the resource name or specifying a name. 

So we'd end up with methods like the following on the relevant container-based resources:

- `WithDataVolume(string? name = null)`
- `WithLogsVolume(string? name = null)`
- `WithDataBindMount(string source)`
- `WithLogsBindMount(string source)`
- `WithInitBindMount(string source)`

**Comment by DamianEdwards** [comment]
@stbau04 are you able to update your approach here to what I [described above](https://github.com/dotnet/aspire/pull/1317#issuecomment-1981890461)?

**Comment by stbau04** [comment]
@DamianEdwards Yes sure

**Comment by stbau04** [comment]
Is that the desired approach we should follow?

**Comment by DamianEdwards** [comment]
> Is that the desired approach we should follow?

That is what I'm currently proposing, Would be good to explicitly hear from @mitchdenny if he has any concerns with that approach.

**Comment by stbau04** [comment]
@DamianEdwards @mitchdenny I've applied the proposed approach to the first few containers. If you are fine with this solution i would continue implementing it for the remaining resources. Do you have any imput for me?

File: src/Aspire.Hosting/MongoDB/MongoDBBuilderExtensions.cs [hunk 2]
```diff
@@ -73,6 +73,41 @@ public static IResourceBuilder<MongoDBDatabaseResource> AddDatabase(this IResour
...
+    /// <summary>
+    /// Adds a named volume for the data folder to a MongoDb container resource.
+    /// </summary>
+    /// <param name="builder">The resource builder.</param>
+    /// <param name="name">The name of the mount. Defaults to an auto-generated volume. </param>
+    /// <param name="isReadOnly">A flag that indicates if this is a read-only mount.</param>
+    /// <returns>The <see cref="IResourceBuilder{T}"/>.</returns>
+    public static IResourceBuilder<MongoDBServerResource> WithDataVolume(this IResourceBuilder<MongoDBServerResource> builder, string? name = null, bool isReadOnly = false)
+    {
+        name ??= $".data/{builder.Resource.Name}";
```
Comment (DamianEdwards) [inline src/Aspire.Hosting/MongoDB/MongoDBBuilderExtensions.cs]: When auto-naming volumes (rather than directories) I think we should use a form more like `"${builder.Resource.Name}-data"`.

File: src/Aspire.Hosting/MongoDB/MongoDBBuilderExtensions.cs [hunk 3]
```diff
@@ -73,6 +73,41 @@ public static IResourceBuilder<MongoDBDatabaseResource> AddDatabase(this IResour
         return builder;
     }
 
+    /// <summary>
+    /// Adds a named volume for the data folder to a MongoDb container resource.
+    /// </summary>
+    /// <param name="builder">The resource builder.</param>
+    /// <param name="name">The name of the mount. Defaults to an auto-generated volume. </param>
```
Comment (DamianEdwards) [inline src/Aspire.Hosting/MongoDB/MongoDBBuilderExtensions.cs:80]:
```suggestion
    /// <param name="name">The name of the volume. Defaults to an auto-generated name based on the resource name. </param>
```

File: src/Aspire.Hosting/MongoDB/MongoDBBuilderExtensions.cs [hunk 4]
```diff
@@ -73,6 +73,41 @@ public static IResourceBuilder<MongoDBDatabaseResource> AddDatabase(this IResour
...
+    /// <param name="isReadOnly">A flag that indicates if this is a read-only mount.</param>
```
Comment (DamianEdwards) [inline src/Aspire.Hosting/MongoDB/MongoDBBuilderExtensions.cs:81]:
```suggestion
    /// <param name="isReadOnly">A flag that indicates if this is a read-only volume.</param>
```

File: src/Aspire.Hosting/MongoDB/MongoDBBuilderExtensions.cs [hunk 5]
```diff
@@ -73,6 +73,41 @@ public static IResourceBuilder<MongoDBDatabaseResource> AddDatabase(this IResour
...
+    /// <returns>The <see cref="IResourceBuilder{T}"/>.</returns>
+    public static IResourceBuilder<MongoDBServerResource> WithDataVolume(this IResourceBuilder<MongoDBServerResource> builder, string? name = null, bool isReadOnly = false)
+    {
+        name ??= $".data/{builder.Resource.Name}";
+        var fullyQualifiedPath = Path.GetFullPath(name, builder.ApplicationBuilder.AppHostDirectory);
+        return builder.WithVolume(fullyQualifiedPath, "/data/db", isReadOnly);
```
Comment (DamianEdwards) [inline src/Aspire.Hosting/MongoDB/MongoDBBuilderExtensions.cs]: The first argument should be the name, not a directory path.

File: src/Aspire.Hosting/MongoDB/MongoDBBuilderExtensions.cs [hunk 6]
```diff
@@ -73,6 +73,41 @@ public static IResourceBuilder<MongoDBDatabaseResource> AddDatabase(this IResour
...
+    }
+
+    /// <summary>
+    /// Adds a bind mount for the data folder to a MongoDb container resource.
+    /// </summary>
+    /// <param name="builder">The resource builder.</param>
+    /// <param name="source">The name of the mount. This is the physical location on the host.</param>
```
Comment (DamianEdwards) [inline src/Aspire.Hosting/MongoDB/MongoDBBuilderExtensions.cs:90]:
```suggestion
    /// <param name="source">The source directory on the host to mount into the container.</param>
```

File: src/Aspire.Hosting/MongoDB/MongoDBBuilderExtensions.cs [hunk 7]
```diff
@@ -73,6 +73,41 @@ public static IResourceBuilder<MongoDBDatabaseResource> AddDatabase(this IResour
...
+    /// <returns>The <see cref="IResourceBuilder{T}"/>.</returns>
+    public static IResourceBuilder<MongoDBServerResource> WithDataBindMount(this IResourceBuilder<MongoDBServerResource> builder, string source, bool isReadOnly = false)
+        => builder.WithBindMount(source, "/data/db", isReadOnly);
+
+    /// <summary>
+    /// Adds a bind mount for the init folder to a MongoDb container resource.
+    /// </summary>
+    /// <typeparam name="T">The resource type.</typeparam>
+    /// <param name="builder">The resource builder.</param>
+    /// <param name="source">The name of the mount. This is the physical location on the host.</param>
```
Comment (DamianEdwards) [inline src/Aspire.Hosting/MongoDB/MongoDBBuilderExtensions.cs:101]: Apply same suggestions regarding doc comments from method above to the rest.

File: src/Aspire.Hosting/MySql/MySqlBuilderExtensions.cs [hunk 8]
```diff
@@ -101,4 +101,39 @@ public static IResourceBuilder<MySqlServerResource> PublishAsContainer(this IRes
...
+    /// </summary>
+    /// <param name="builder">The resource builder.</param>
+    /// <param name="name">The name of the mount. Defaults to an auto-generated volume. </param>
+    /// <param name="isReadOnly">A flag that indicates if this is a read-only mount.</param>
+    /// <returns>The <see cref="IResourceBuilder{T}"/>.</returns>
+    public static IResourceBuilder<MySqlServerResource> WithDataVolume(this IResourceBuilder<MySqlServerResource> builder, string? name = null, bool isReadOnly = false)
+    {
+        name ??= $".data/{builder.Resource.Name}";
+        var fullyQualifiedPath = Path.GetFullPath(name, builder.ApplicationBuilder.AppHostDirectory);
+        return builder.WithVolume(fullyQualifiedPath, "/var/lib/mysql", isReadOnly);
```
Comment (DamianEdwards) [inline src/Aspire.Hosting/MySql/MySqlBuilderExtensions.cs]: Same as above, first parameter is the volume name.

File: src/Aspire.Hosting/Postgres/PostgresBuilderExtensions.cs [hunk 9]
```diff
@@ -115,4 +115,39 @@ public static IResourceBuilder<PostgresServerResource> PublishAsContainer(this I
...
+    /// </summary>
+    /// <param name="builder">The resource builder.</param>
+    /// <param name="name">The name of the mount. Defaults to an auto-generated volume. </param>
+    /// <param name="isReadOnly">A flag that indicates if this is a read-only mount.</param>
+    /// <returns>The <see cref="IResourceBuilder{T}"/>.</returns>
+    public static IResourceBuilder<PostgresServerResource> WithDataVolume(this IResourceBuilder<PostgresServerResource> builder, string? name = null, bool isReadOnly = false)
+    {
+        name ??= $".data/{builder.Resource.Name}";
+        var fullyQualifiedPath = Path.GetFullPath(name, builder.ApplicationBuilder.AppHostDirectory);
+        return builder.WithVolume(fullyQualifiedPath, "/var/lib/postgresql/data", isReadOnly);
```
Comment (DamianEdwards) [inline src/Aspire.Hosting/Postgres/PostgresBuilderExtensions.cs]: First parameter is the volume name.

File: src/Aspire.Hosting/RabbitMQ/RabbitMQBuilderExtensions.cs [hunk 10]
```diff
@@ -51,4 +51,28 @@ public static IResourceBuilder<RabbitMQServerResource> PublishAsContainer(this I
...
+    /// </summary>
+    /// <param name="builder">The resource builder.</param>
+    /// <param name="name">The name of the mount. Defaults to an auto-generated volume. </param>
+    /// <param name="isReadOnly">A flag that indicates if this is a read-only mount.</param>
+    /// <returns>The <see cref="IResourceBuilder{T}"/>.</returns>
+    public static IResourceBuilder<RabbitMQServerResource> WithDataVolume(this IResourceBuilder<RabbitMQServerResource> builder, string? name = null, bool isReadOnly = false)
+    {
+        name ??= $".data/{builder.Resource.Name}";
+        var fullyQualifiedPath = Path.GetFullPath(name, builder.ApplicationBuilder.AppHostDirectory);
+        return builder.WithVolume(fullyQualifiedPath, "/var/lib/rabbitmq", isReadOnly);
```
Comment (DamianEdwards) [inline src/Aspire.Hosting/RabbitMQ/RabbitMQBuilderExtensions.cs]: First parameter is the volume name.

File: src/Aspire.Hosting/Redis/RedisBuilderExtensions.cs [hunk 11]
```diff
@@ -54,4 +54,28 @@ public static IResourceBuilder<RedisResource> WithRedisCommander(this IResourceB
...
+    /// </summary>
+    /// <param name="builder">The resource builder.</param>
+    /// <param name="name">The name of the mount. Defaults to an auto-generated volume. </param>
+    /// <param name="isReadOnly">A flag that indicates if this is a read-only mount.</param>
+    /// <returns>The <see cref="IResourceBuilder{T}"/>.</returns>
+    public static IResourceBuilder<RedisResource> WithDataVolume(this IResourceBuilder<RedisResource> builder, string? name = null, bool isReadOnly = false)
+    {
+        name ??= $".data/{builder.Resource.Name}";
+        var fullyQualifiedPath = Path.GetFullPath(name, builder.ApplicationBuilder.AppHostDirectory);
+        return builder.WithVolume(fullyQualifiedPath, "/data", isReadOnly);
```
Comment (DamianEdwards) [inline src/Aspire.Hosting/Redis/RedisBuilderExtensions.cs]: First parameter is the volume name.

File: src/Aspire.Hosting/SqlServer/SqlServerBuilderExtensions.cs [hunk 12]
```diff
@@ -72,4 +72,52 @@ public static IResourceBuilder<SqlServerDatabaseResource> AddDatabase(this IReso
...
+    /// </summary>
+    /// <param name="builder">The resource builder.</param>
+    /// <param name="name">The name of the mount. Defaults to an auto-generated volume. </param>
+    /// <param name="isReadOnly">A flag that indicates if this is a read-only mount.</param>
+    /// <returns>The <see cref="IResourceBuilder{T}"/>.</returns>
+    public static IResourceBuilder<SqlServerServerResource> WithDataVolume(this IResourceBuilder<SqlServerServerResource> builder, string? name = null, bool isReadOnly = false)
+    {
+        name ??= $".data/{builder.Resource.Name}";
+        var fullyQualifiedPath = Path.GetFullPath(name, builder.ApplicationBuilder.AppHostDirectory);
+        return builder.WithVolume(fullyQualifiedPath, "/var/opt/mssql/data", isReadOnly);
```
Comment (DamianEdwards) [inline src/Aspire.Hosting/SqlServer/SqlServerBuilderExtensions.cs]: First parameter is the volume name.

**Review by DamianEdwards — CHANGES_REQUESTED** [review]

File: src/Aspire.Hosting/MongoDB/MongoDBBuilderExtensions.cs (hunk 2 above)
Comment (stbau04) [reply src/Aspire.Hosting/MongoDB/MongoDBBuilderExtensions.cs]: I know, this is already fixed (the second commit i pushed today). The pipeline for the fix failed on the "Run helix tests" task, doesn't seem like a problem with my changes (https://dev.azure.com/dnceng-public/public/_build/results?buildId=596222&view=logs&jobId=0bc77094-9fcd-5c38-f6e4-27d2ae131589&j=0bc77094-9fcd-5c38-f6e4-27d2ae131589&t=06901622-1165-5cb8-9fdb-a339481d3deb)

**Comment by DamianEdwards** [comment]
@stbau04 thanks for the updates! Were there more resource types you were planning to apply this pattern to?

**Comment by davidfowl** [comment]
I would leave the azure resources as they are currently being renovated by @mitchdenny. @mitchdenny you can take this pattern to the existing emulators.

**Comment by stbau04** [comment]
@DamianEdwards I could maybe add some methods for the following: KafkaServerResource, MongoExpressContainerResource, PhpMyAdminContainerResource, OracleDatabaseServerResource, PgAdminContainerResource and RedisCommandResource (might take me some time to figure out the correct paths)
If this is not necessary/wantend at the moment it would be ready to review

**Comment by davidfowl** [comment]
KafkaServerResource - 👍🏾 
OracleDatabaseServerResource - 👍🏾 

The rest I would skip.

**Comment by stbau04** [comment]
@DamianEdwards I've added some method for the KafkaServer and OracleDatabaseServer resoures. I think it should be ready for review now

**Review by DamianEdwards — APPROVED** [review]

**Comment by davidfowl** [comment]
@mitchdenny We need to change cosmos and azure storage to use the same pattern yes? `WithVolumeForData`?

**Comment by davidfowl** [comment]
Seq differs a little here:

https://github.com/dotnet/aspire/blob/c236e90636e6b0b4bba7a28e7dc60ab5a50f45c0/src/Aspire.Hosting/Seq/SeqBuilderExtensions.cs#L27

We should make this follow the same pattern I think.

**Comment by davidfowl** [comment]
It can be a follow up though

**Comment by davidfowl** [comment]
Thanks @stbau04 !!!!
